[SPARK-4176][WIP] Support decimal types with precision > 18 in parquet

rtreffer opened the pull request:
This is my current WIP on SPARK-4176. It should be compatible with other implementations of parquet.

https://github.com/Parquet/parquet-format/blob/master/LogicalTypes.md#decimal

```
For byte arrays, binary and fixed, the unscaled number must be encoded as
two's complement using big-endian byte order
(the most significant byte is the zeroth element)
```

This is the default encoding on bigint. It should thus be compatible with other implementations, although it would be great if s.o. could test this.

I've tested this locally with powers of 2 up to 2^200 in the spark shell, without errors but

Code I've used for (local) testing (on spark shell):

``` scala
import org.apache.spark.sql.types._
import org.apache.spark.sql.Row
import java.io.File

def decimalList(n : Int) = {
  require(n > 0)
  val one = Decimal(1)
  val two = Decimal(2)
  one.changePrecision(n,0)
  two.changePrecision(n,0)
  val list = scala.collection.mutable.ArrayBuffer(one)
  while (list.length < n) {
    val v = list(list.length - 1)
    list += Decimal(v.toJavaBigDecimal.multiply(two.toJavaBigDecimal),n,0)
  }
  list.toList
}

def decimalRows(l : List[Decimal]) = l.zipWithIndex.map(e => Row(e._2,e._1))

def decimalRowRdd(l : List[Decimal]) = sc.parallelize(decimalRows(l))

def df(n : Int) = {
  val data = decimalList(n)
  val len = data.lastOption.get.toString.length
  val schema = StructType(Array(
    StructField("id", IntegerType, true),
    StructField("value", DecimalType(len,0), true)))
  sqlContext.createDataFrame(decimalRowRdd(data), schema)
}

def delete(filename : String) : Unit = {
  val f = new File(filename)
  if (!f.exists) return
  if (f.isDirectory) f.listFiles.foreach(_.delete)
  f.delete
}

def test(n : Int) = {
  val src = df(n)
  delete("/tmp/typetest")
  src.save("/tmp/typetest")
  val copy = sqlContext.load("/tmp/typetest")
  src.collect().sortBy(_.getInt(0)).zip(copy.collect().sortBy(_.getInt(0))).foreach(e => {
    if (e._1 != e._2) println(s"${e._1} != ${e._2}")
  })
  delete("/tmp/typetest")
  println(s"Tested 1..2^${n - 1}")
}
```

AmplabJenkins commented:
Can one of the admins verify this patch?

rtreffer commented:
Note: I came across https://issues.apache.org/jira/browse/SPARK-8342 while testing, it seems like Decimal math is unsafe at the moment.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala:
```diff
@@ -331,7 +331,7 @@ private[sql] class JDBCRDD(
       case DateType => DateConversion
       case DecimalType.Unlimited => DecimalConversion(None)
-      case DecimalType.Fixed(d) => DecimalConversion(Some(d))
+      case DecimalType.Fixed(d, s) => DecimalConversion(Some((d, s)))
```
The type of `d` is `Option[(Int, Int)]`. So I think it already contains precision and scale, why to add `s`?

rtreffer replied:
It caused a warning on my system. So I thought it would be better to make it explicit.

I could drop that line from this patch, though.

liancheng replied:
This should be correct. Scala pattern extractors use tuples if they want to return multiple values.

rtreffer replied:
As said it was only about a warning, not about correctness. I'll drop this change on the next version, it draws too much attention and is not needed.

marmbrus commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35062 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35062/consoleFull) for   PR 6796 at commit [`8f6445c`](https://github.com/apache/spark/commit/8f6445c25fa62cd9fa3cb28b7441dd19d8692c6b).

SparkQA commented:
[Test build #35062 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35062/console) for   PR 6796 at commit [`8f6445c`](https://github.com/apache/spark/commit/8f6445c25fa62cd9fa3cb28b7441dd19d8692c6b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35115 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35115/consoleFull) for   PR 6796 at commit [`7310902`](https://github.com/apache/spark/commit/7310902b8416af58ff1dc4ad6dbe65fcb76fd520).

SparkQA commented:
[Test build #35115 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35115/console) for   PR 6796 at commit [`7310902`](https://github.com/apache/spark/commit/7310902b8416af58ff1dc4ad6dbe65fcb76fd520).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

marmbrus commented:
Thanks for working on this! Did a quick pass and it looks pretty good.  I'll let @liancheng do a more complete review.  Does SPARK-8342 block merging this or can you remove the WIP tag now?

rtreffer commented:
The only reason for the WIP is that I have not yet cross-tested the interoperability with e.g. hive. It follows the spec, but I'd like to test it (or have s.o. verify this).

marmbrus commented:
Ah cool, it might even be a good idea to check in small files that are created with other systems for these kinds of tests.

rtreffer commented:
I was thinking about this: Create a small parquet file with spark, load it with hive, copy it to a new parquet table with hive, read that with spark.

If that matches the input -> win. Otherwise -> some more work.

PS: SPARK-8342 / SPARK-8359 were only problems during my initial tests. It's a bit harder to test read/write of Decimal if the implementation has bugs. So those are unrelated to this patch, but they might reduce the usefulness of this patch (you can't do reliable math in the ranges you could now load/save)

marmbrus commented:
Are you thinking about doing this as part of the test or doing it manually?  Right now parquet and its tests have no hive dependencies, which I think is good.  But I would definitely like to have a test that reads a file that was written by hive.

rtreffer commented:
Yes, manually. I could add the file I was writing afterwards, sounds like a good idea.

rtreffer commented:
Just did a test with hive, I can declare a parquet file written with spark as

{code}
CREATE EXTERNAL TABLE ptest (id INT, value DECIMAL(30,0))
STORED AS PARQUET
LOCATION "file:///home/rtreffer/work/hadoop/parquet";
{code}

and it *does* work. I'm now trying to test the opposite direction plus a test case.

rtreffer commented:
(hive 1.2.0 and hadoop 2.7.0 without hdfs or a cluster)

rtreffer commented:
Ok, it looks like I can't open hive generated parquet files, but it looks more like a type error.

```
scala> val hive = sqlContext.load("/home/rtreffer/work/hadoop/hive-parquet")
warning: there was one deprecation warning; re-run with -deprecation for details
15/06/19 22:03:26 INFO ParquetFileReader: Initiating action with parallelism: 5
hive: org.apache.spark.sql.DataFrame = [id: int, value: decimal(30,0)]

scala> hive.collect.foreach(println)
15/06/19 22:03:35 INFO BlockManagerInfo: Removed broadcast_8_piece0 on localhost:42189 in memory (size: 2.4 KB, free: 265.1 MB)
15/06/19 22:03:35 INFO BlockManagerInfo: Removed broadcast_9_piece0 on localhost:42189 in memory (size: 2.4 KB, free: 265.1 MB)
15/06/19 22:03:35 INFO MemoryStore: ensureFreeSpace(130208) called with curMem=0, maxMem=278019440
15/06/19 22:03:35 INFO MemoryStore: Block broadcast_10 stored as values in memory (estimated size 127.2 KB, free 265.0 MB)
15/06/19 22:03:35 INFO MemoryStore: ensureFreeSpace(14082) called with curMem=130208, maxMem=278019440
15/06/19 22:03:35 INFO MemoryStore: Block broadcast_10_piece0 stored as bytes in memory (estimated size 13.8 KB, free 265.0 MB)
15/06/19 22:03:35 INFO BlockManagerInfo: Added broadcast_10_piece0 in memory on localhost:42189 (size: 13.8 KB, free: 265.1 MB)
15/06/19 22:03:35 INFO SparkContext: Created broadcast 10 from collect at <console>:38
15/06/19 22:03:35 INFO SparkContext: Starting job: collect at <console>:38
15/06/19 22:03:35 INFO DAGScheduler: Got job 9 (collect at <console>:38) with 1 output partitions (allowLocal=false)
15/06/19 22:03:35 INFO DAGScheduler: Final stage: ResultStage 11(collect at <console>:38)
15/06/19 22:03:35 INFO DAGScheduler: Parents of final stage: List()
15/06/19 22:03:35 INFO DAGScheduler: Missing parents: List()
15/06/19 22:03:35 INFO DAGScheduler: Submitting ResultStage 11 (MapPartitionsRDD[45] at collect at <console>:38), which has no missing parents
15/06/19 22:03:35 INFO MemoryStore: ensureFreeSpace(5568) called with curMem=144290, maxMem=278019440
15/06/19 22:03:35 INFO MemoryStore: Block broadcast_11 stored as values in memory (estimated size 5.4 KB, free 265.0 MB)
15/06/19 22:03:35 INFO MemoryStore: ensureFreeSpace(2964) called with curMem=149858, maxMem=278019440
15/06/19 22:03:35 INFO MemoryStore: Block broadcast_11_piece0 stored as bytes in memory (estimated size 2.9 KB, free 265.0 MB)
15/06/19 22:03:36 INFO BlockManagerInfo: Added broadcast_11_piece0 in memory on localhost:42189 (size: 2.9 KB, free: 265.1 MB)
15/06/19 22:03:36 INFO SparkContext: Created broadcast 11 from broadcast at DAGScheduler.scala:893
15/06/19 22:03:36 INFO DAGScheduler: Submitting 1 missing tasks from ResultStage 11 (MapPartitionsRDD[45] at collect at <console>:38)
15/06/19 22:03:36 INFO TaskSchedulerImpl: Adding task set 11.0 with 1 tasks
15/06/19 22:03:36 INFO TaskSetManager: Starting task 0.0 in stage 11.0 (TID 36, localhost, PROCESS_LOCAL, 1471 bytes)
15/06/19 22:03:36 INFO Executor: Running task 0.0 in stage 11.0 (TID 36)
15/06/19 22:03:36 INFO ParquetRelation2$$anonfun$buildScan$1$$anon$1: Input split: ParquetInputSplit{part: file:/home/rtreffer/work/hadoop/hive-parquet/000000_0 start: 0 end: 874 length: 874 hosts: []}
15/06/19 22:03:36 INFO InternalParquetRecordReader: RecordReader initialized will read a total of 33 records.
15/06/19 22:03:36 INFO InternalParquetRecordReader: at row 0. reading next block
15/06/19 22:03:36 INFO InternalParquetRecordReader: block read in memory in 0 ms. row count = 33
15/06/19 22:03:36 ERROR Executor: Exception in task 0.0 in stage 11.0 (TID 36)
org.apache.parquet.io.ParquetDecodingException: Can not read value at 0 in block -1 in file file:/home/rtreffer/work/hadoop/hive-parquet/000000_0
        at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:228)
        at org.apache.parquet.hadoop.ParquetRecordReader.nextKeyValue(ParquetRecordReader.java:201)
        at org.apache.spark.sql.sources.SqlNewHadoopRDD$$anon$1.hasNext(SqlNewHadoopRDD.scala:163)
        at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:39)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$class.foreach(Iterator.scala:750)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1202)
        at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:59)
        at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:104)
        at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:48)
        at scala.collection.TraversableOnce$class.to(TraversableOnce.scala:295)
        at scala.collection.AbstractIterator.to(Iterator.scala:1202)
        at scala.collection.TraversableOnce$class.toBuffer(TraversableOnce.scala:287)
        at scala.collection.AbstractIterator.toBuffer(Iterator.scala:1202)
        at scala.collection.TraversableOnce$class.toArray(TraversableOnce.scala:274)
        at scala.collection.AbstractIterator.toArray(Iterator.scala:1202)
        at org.apache.spark.rdd.RDD$$anonfun$collect$1$$anonfun$12.apply(RDD.scala:885)
        at org.apache.spark.rdd.RDD$$anonfun$collect$1$$anonfun$12.apply(RDD.scala:885)
        at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1765)
        at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1765)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:63)
        at org.apache.spark.scheduler.Task.run(Task.scala:70)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.parquet.io.ParquetDecodingException: The requested schema is not compatible with the file schema. incompatible types: optional binary value (DECIMAL(30,0)) != optional fixed_len_byte_array(13) value (DECIMAL(30,0))
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.incompatibleSchema(ColumnIOFactory.java:106)
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.visit(ColumnIOFactory.java:98)
        at org.apache.parquet.schema.PrimitiveType.accept(PrimitiveType.java:389)
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.visitChildren(ColumnIOFactory.java:88)
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.visit(ColumnIOFactory.java:62)
        at org.apache.parquet.schema.MessageType.accept(MessageType.java:58)
        at org.apache.parquet.io.ColumnIOFactory.getColumnIO(ColumnIOFactory.java:149)
        at org.apache.parquet.hadoop.InternalParquetRecordReader.checkRead(InternalParquetRecordReader.java:136)
        at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:208)
        ... 28 more
15/06/19 22:03:36 WARN TaskSetManager: Lost task 0.0 in stage 11.0 (TID 36, localhost): org.apache.parquet.io.ParquetDecodingException: Can not read value at 0 in block -1 in file file:/home/rtreffer/work/hadoop/hive-parquet/000000_0
        at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:228)
        at org.apache.parquet.hadoop.ParquetRecordReader.nextKeyValue(ParquetRecordReader.java:201)
        at org.apache.spark.sql.sources.SqlNewHadoopRDD$$anon$1.hasNext(SqlNewHadoopRDD.scala:163)
        at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:39)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$class.foreach(Iterator.scala:750)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1202)
        at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:59)
        at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:104)
        at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:48)
        at scala.collection.TraversableOnce$class.to(TraversableOnce.scala:295)
        at scala.collection.AbstractIterator.to(Iterator.scala:1202)
        at scala.collection.TraversableOnce$class.toBuffer(TraversableOnce.scala:287)
        at scala.collection.AbstractIterator.toBuffer(Iterator.scala:1202)
        at scala.collection.TraversableOnce$class.toArray(TraversableOnce.scala:274)
        at scala.collection.AbstractIterator.toArray(Iterator.scala:1202)
        at org.apache.spark.rdd.RDD$$anonfun$collect$1$$anonfun$12.apply(RDD.scala:885)
        at org.apache.spark.rdd.RDD$$anonfun$collect$1$$anonfun$12.apply(RDD.scala:885)
        at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1765)
        at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1765)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:63)
        at org.apache.spark.scheduler.Task.run(Task.scala:70)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.parquet.io.ParquetDecodingException: The requested schema is not compatible with the file schema. incompatible types: optional binary value (DECIMAL(30,0)) != optional fixed_len_byte_array(13) value (DECIMAL(30,0))
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.incompatibleSchema(ColumnIOFactory.java:106)
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.visit(ColumnIOFactory.java:98)
        at org.apache.parquet.schema.PrimitiveType.accept(PrimitiveType.java:389)
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.visitChildren(ColumnIOFactory.java:88)
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.visit(ColumnIOFactory.java:62)
        at org.apache.parquet.schema.MessageType.accept(MessageType.java:58)
        at org.apache.parquet.io.ColumnIOFactory.getColumnIO(ColumnIOFactory.java:149)
        at org.apache.parquet.hadoop.InternalParquetRecordReader.checkRead(InternalParquetRecordReader.java:136)
        at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:208)
        ... 28 more

15/06/19 22:03:36 ERROR TaskSetManager: Task 0 in stage 11.0 failed 1 times; aborting job
15/06/19 22:03:36 INFO TaskSchedulerImpl: Removed TaskSet 11.0, whose tasks have all completed, from pool
15/06/19 22:03:36 INFO TaskSchedulerImpl: Cancelling stage 11
15/06/19 22:03:36 INFO DAGScheduler: ResultStage 11 (collect at <console>:38) failed in 0.021 s
15/06/19 22:03:36 INFO DAGScheduler: Job 9 failed: collect at <console>:38, took 0.033099 s
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 11.0 failed 1 times, most recent failure: Lost task 0.0 in stage 11.0 (TID 36, localhost): org.apache.parquet.io.ParquetDecodingException: Can not read value at 0 in block -1 in file file:/home/rtreffer/work/hadoop/hive-parquet/000000_0
        at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:228)
        at org.apache.parquet.hadoop.ParquetRecordReader.nextKeyValue(ParquetRecordReader.java:201)
        at org.apache.spark.sql.sources.SqlNewHadoopRDD$$anon$1.hasNext(SqlNewHadoopRDD.scala:163)
        at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:39)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
        at scala.collection.Iterator$class.foreach(Iterator.scala:750)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1202)
        at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:59)
        at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:104)
        at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:48)
        at scala.collection.TraversableOnce$class.to(TraversableOnce.scala:295)
        at scala.collection.AbstractIterator.to(Iterator.scala:1202)
        at scala.collection.TraversableOnce$class.toBuffer(TraversableOnce.scala:287)
        at scala.collection.AbstractIterator.toBuffer(Iterator.scala:1202)
        at scala.collection.TraversableOnce$class.toArray(TraversableOnce.scala:274)
        at scala.collection.AbstractIterator.toArray(Iterator.scala:1202)
        at org.apache.spark.rdd.RDD$$anonfun$collect$1$$anonfun$12.apply(RDD.scala:885)
        at org.apache.spark.rdd.RDD$$anonfun$collect$1$$anonfun$12.apply(RDD.scala:885)
        at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1765)
        at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1765)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:63)
        at org.apache.spark.scheduler.Task.run(Task.scala:70)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.parquet.io.ParquetDecodingException: The requested schema is not compatible with the file schema. incompatible types: optional binary value (DECIMAL(30,0)) != optional fixed_len_byte_array(13) value (DECIMAL(30,0))
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.incompatibleSchema(ColumnIOFactory.java:106)
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.visit(ColumnIOFactory.java:98)
        at org.apache.parquet.schema.PrimitiveType.accept(PrimitiveType.java:389)
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.visitChildren(ColumnIOFactory.java:88)
        at org.apache.parquet.io.ColumnIOFactory$ColumnIOCreatorVisitor.visit(ColumnIOFactory.java:62)
        at org.apache.parquet.schema.MessageType.accept(MessageType.java:58)
        at org.apache.parquet.io.ColumnIOFactory.getColumnIO(ColumnIOFactory.java:149)
        at org.apache.parquet.hadoop.InternalParquetRecordReader.checkRead(InternalParquetRecordReader.java:136)
        at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:208)
        ... 28 more

Driver stacktrace:
  at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1285)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1276)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1275)
  at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
  at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
  at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1275)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:749)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:749)
  at scala.Option.foreach(Option.scala:257)
  at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:749)
  at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1484)
  at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1445)
  at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
```

Hm, could be that the spark decoder is too strict. There are various ways to encode DECIMAL(30), and it looks like hive chooses fixed_len arrays, while I prefer variable length arrays. Have to double check that.

rtreffer commented:
I've pushed the hive generated parquet file and I'll call it a day.

I think I'll have to relax the validation of column types for DECIMAL.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35317 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35317/consoleFull) for   PR 6796 at commit [`b8f2f2d`](https://github.com/apache/spark/commit/b8f2f2d834b2575f847438e77fc735f5ad0a72e7).

SparkQA commented:
[Test build #35317 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35317/console) for   PR 6796 at commit [`b8f2f2d`](https://github.com/apache/spark/commit/b8f2f2d834b2575f847438e77fc735f5ad0a72e7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

rtreffer commented:
Ok, I think I'm slowly getting down to the cause....
strictTypeChecking on https://github.com/Parquet/parquet-mr/blob/master/parquet-column/src/main/java/parquet/io/ColumnIOFactory.java#L100 creates the exception

The relevant class for the job setup is ParquetTableScan (doExecute).

I'm not yet sure if this can be fixed on the job setup or on the receiver side.

rtreffer commented:
The problematic line is
https://github.com/apache/spark/blob/v1.4.0/sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableSupport.scala#L105
(calling ParquetTypesConverter.convertFromAttributes)
This causes the schema to be normalized, but that does not work for decimal.

```
message hive_schema {
  optional int32 id;
  optional fixed_len_byte_array(13) value (DECIMAL(30,0));
}
```
is replaced by
```
message root {
  optional int32 id;
  optional binary value (DECIMAL(30,0));
}
```

I removed that line and the loading of the data works.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

rtreffer commented:
I've pushed a very early version of a fix. (Literally early, it's nearly 1:00 am. And I'd expect the test build to fail, I'll fix the outstanding issues later today)

PS: Loading the hive parquet works now, but I've not yet tested much more.

SparkQA commented:
[Test build #35373 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35373/consoleFull) for   PR 6796 at commit [`464d24e`](https://github.com/apache/spark/commit/464d24ea68fb8be301c683cc3eccfc900c2d18bb).

davies commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala:
```diff
@@ -289,7 +295,20 @@ private[parquet] object ParquetTypesConverter extends Logging {
+      parquetSchema: Option[ParquetType] = None,
       toThriftSchemaNames: Boolean = false): ParquetType = {
+
+    val parquetElementTypeBySchema =
```
These could be
```
parquetSchema.filter(_.containsField(name)).map(_.getType(name))
```

rtreffer replied:
Ah, yes, an early version had that, I somehow moved to this verbose code O.o

Thanks

Am 21. Juni 2015 02:33:15 MESZ, schrieb Davies Liu <notifications@github.com>:
>>        toThriftSchemaNames: Boolean = false): ParquetType = {
>> +
>> +    val parquetElementTypeBySchema =
>
>These could be
>```
>parquetSchema.filter(_.containsField(name)).map(_.getType(name))
>```
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/apache/spark/pull/6796/files#r32889401

--

rtreffer replied:
It also performs a type check / conversion. That's why I've removed it. It would look like this

```
val parquetElementTypeBySchema =
  parquetSchema.filter(_.isInstanceOf[ParquetGroupType]).filter(_.containsField(name)).map(_.getType(name))
```

I would settle on collect, does that look ok?

```
  val parquetElementTypeBySchema = parquetSchema.collect {
    case gType : ParquetGroupType if (gType.containsField(name)) => gType.getType(name)
  }
```

davies replied:
This one is better

SparkQA commented:
[Test build #35373 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35373/console) for   PR 6796 at commit [`464d24e`](https://github.com/apache/spark/commit/464d24ea68fb8be301c683cc3eccfc900c2d18bb).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

davies commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala:
```diff
@@ -229,11 +231,15 @@ private[parquet] object ParquetTypesConverter extends Logging {
     case TimestampType => Some(ParquetTypeInfo(ParquetPrimitiveTypeName.INT96))
     case DecimalType.Fixed(precision, scale) if precision <= 18 =>
-      // TODO: for now, our writer only supports decimals that fit in a Long
       Some(ParquetTypeInfo(ParquetPrimitiveTypeName.FIXED_LEN_BYTE_ARRAY,
```
We should int32, int64 if possible, see https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#decimal
```
DECIMAL can be used to annotate the following types:

int32: for 1 <= precision <= 9
int64: for 1 <= precision <= 18; precision <= 10 will produce a warning
fixed_len_byte_array: precision is limited by the array size. Length n can store <= floor(log_10(2^(8*n - 1) - 1)) base-10 digits
binary: precision is not limited, but is required. The minimum number of bytes to store the unscaled value should be used.
```

rtreffer replied:
What would we gain by encoding it that way?
We use a minimal length fixed byte array which should provide a similar compact encoding. (DECIMAL(9) should end up as 4 bytes, and smaller decimal values should take even less space)

Decoding is a different story, though.

PS: I was focusing on DECIMAL with precision >=19. Shouldn't small decimal handling be a new ticket?

Am 21. Juni 2015 02:41:44 MESZ, schrieb Davies Liu <notifications@github.com>:
>> @@ -229,11 +231,15 @@ private[parquet] object ParquetTypesConverter
>extends Logging {
>>      case LongType =>
>Some(ParquetTypeInfo(ParquetPrimitiveTypeName.INT64))
>>      case TimestampType =>
>Some(ParquetTypeInfo(ParquetPrimitiveTypeName.INT96))
>>      case DecimalType.Fixed(precision, scale) if precision <= 18 =>
>> -      // TODO: for now, our writer only supports decimals that fit
>in a Long
>>       
>Some(ParquetTypeInfo(ParquetPrimitiveTypeName.FIXED_LEN_BYTE_ARRAY,
>
>We should int32, int64 if possible, see
>https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#decimal
>```
>DECIMAL can be used to annotate the following types:
>
>int32: for 1 <= precision <= 9
>int64: for 1 <= precision <= 18; precision <= 10 will produce a warning
>fixed_len_byte_array: precision is limited by the array size. Length n
>can store <= floor(log_10(2^(8*n - 1) - 1)) base-10 digits
>binary: precision is not limited, but is required. The minimum number
>of bytes to store the unscaled value should be used.
>```
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/apache/spark/pull/6796/files#r32889441

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

davies replied:
Make sense.

liancheng replied:
Using int32 and int64 makes encoding and decoding faster since they don't introduce boxing costs. But I agree that should be made in another PR.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35414 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35414/consoleFull) for   PR 6796 at commit [`f973b58`](https://github.com/apache/spark/commit/f973b582150f21a8d8e97937585cd11c407f29fc).

SparkQA commented:
[Test build #35414 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35414/console) for   PR 6796 at commit [`f973b58`](https://github.com/apache/spark/commit/f973b582150f21a8d8e97937585cd11c407f29fc).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35545 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35545/consoleFull) for   PR 6796 at commit [`8ff6603`](https://github.com/apache/spark/commit/8ff660369601418c36ed82e9549bd676a83b8345).

SparkQA commented:
[Test build #35545 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35545/console) for   PR 6796 at commit [`8ff6603`](https://github.com/apache/spark/commit/8ff660369601418c36ed82e9549bd676a83b8345).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala:
```diff
@@ -289,7 +295,13 @@ private[parquet] object ParquetTypesConverter extends Logging {
       toThriftSchemaNames: Boolean = false): ParquetType = {
+
+    val parquetElementTypeBySchema = parquetSchema.collect {
+        case gType : ParquetGroupType if (gType.containsField(name)) => gType.getType(name)
```
Nit: Remove the space before `:`

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableSupport.scala:
```diff
@@ -369,9 +371,6 @@ private[parquet] class MutableRowWriteSupport extends RowWriteSupport {
       case d: DecimalType =>
-        if (d.precisionInfo == None || d.precisionInfo.get.precision > 18) {
-          sys.error(s"Unsupported datatype $d, cannot write to consumer")
-        }
```
Don't we need to consider the case where `d.precisionInfo == None` now?

rtreffer replied:
Overseen, bug. We can't serialize without that info, parquet requires it (and mixed scale would be complicated).

PS: do you know if there is any interest on allowing mixed Decimal in parquet?

Am 24. Juni 2015 09:36:11 MESZ, schrieb Cheng Lian <notifications@github.com>:
>> @@ -369,9 +371,6 @@ private[parquet] class MutableRowWriteSupport
>extends RowWriteSupport {
>>        case DateType => writer.addInteger(record.getInt(index))
>>        case TimestampType => writeTimestamp(record.getLong(index))
>>        case d: DecimalType =>
>> -        if (d.precisionInfo == None || d.precisionInfo.get.precision
>> 18) {
>> -          sys.error(s"Unsupported datatype $d, cannot write to
>consumer")
>> -        }
>
>Don't we need to consider the case where `d.precisionInfo == None` now?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/apache/spark/pull/6796/files#r33123740

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

liancheng replied:
I don't think Parquet allow mixed decimal precision. However, Hive uses a default precision = 10 and a default scale = 0 when precision/scale information is missing. I also did the same thing in #6617.

liancheng replied:
Had an offline discussion with @yhuai but forgot to post a summary here: at last we decided not to convert unlimited decimals to `decimal(10, 0)` implicitly in #6617, because firstly we need to confirm all other parts works in a consistent way, which might introduce unexpected complexity in #6617, and secondly implicit conversion can often become a huge footgun. So let's still report an error in case of `d.precisionInfo == None` (but please throw an `AnalysisException` instead of using `sys.error`).

liancheng commented on sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetIOSuite.scala:
```diff
@@ -108,7 +108,7 @@ class ParquetIOSuiteBase extends QueryTest with ParquetTest {
         .select($"_1" cast decimal as "dec")
 
-    for ((precision, scale) <- Seq((5, 2), (1, 0), (1, 1), (18, 10), (18, 17))) {
+    for ((precision, scale) <- Seq((5, 2), (1, 0), (1, 1), (18, 10), (18, 17), (60, 5))) {
```
It would be good to add one more edge case here, namely `(19, n)`.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala:
```diff
@@ -229,11 +231,15 @@ private[parquet] object ParquetTypesConverter extends Logging {
         Some(new DecimalMetadata(precision, scale)),
         Some(BYTES_FOR_PRECISION(precision))))
+    case DecimalType.Fixed(precision, scale) =>
+      Some(ParquetTypeInfo(ParquetPrimitiveTypeName.BINARY,
```
Using `BINARY` here conforms to Parquet format spec. But according to the spec, `FIXED_LENGTH_BYTE_ARRAY` with different length can also be used to store decimals with different precisions. From the perspective of storage efficiency, `FIXED_LENGTH_BYTE_ARRAY` is probably more preferable, since `BINARY` has variable length and needs 4 extra bytes to encode the length (before being encoded and compressed).

Another benefit here is that we can just unify cases for precision <= 18 and precision > 18.

rtreffer replied:
Under the assumption that all values will use the full length, yes.

But at some point the overhead of the length is low compared to the overhead if someone specifies just the upper bound of values.
I have to check if it really uses 4 bytes for BINARY. I'd then raise the threshold to ~40 bytes length. (meaning <=10% worst case overhead before compression)

It won't simplify the decoding/writing though, because the <=18 case  is used for long decoding.

liancheng commented:
@rtreffer I'm working on improving compatibility and interoperability of Spark SQL's Parquet support. The first part is #6617, where I refactored the scheme conversion code so that now we can stick to the most recent Parquet format spec (`ParquetTypes.scala` replaced with `CatalystSchemaConverter.scala`). Schema conversion part of the decimal precision problem is also handled there ([1] [1], [2] [2]). Would you mind if I merge that one and then you rebase this PR? I think it would be much easier to work with. Basically you only need to:

1. Remove the `precision <= ...` part in [this line] [3], and
1. Always use `FIXED_LENGTH_BYTE_ARRAY` to store decimals

[1]: https://github.com/apache/spark/pull/6617/files#diff-a4c01298c63223d113645a31c01141baR370
[2]: https://github.com/apache/spark/pull/6617/files#diff-a4c01298c63223d113645a31c01141baR118
[3]: https://github.com/apache/spark/pull/6617/files#diff-a4c01298c63223d113645a31c01141baR377

rtreffer commented:
@liancheng I'll rebase on your branch, I really like the way you cleaned up toPrimitiveDataType by using a fluent Types interface. This will make this patch way easier.

Talking about testing/compatibility/interoperability, I have added a hive-generated parquet file that I'd like to turn into a test case:
https://github.com/rtreffer/spark/tree/spark-4176-store-large-decimal-in-parquet/sql/core/src/test/resources/hive-decimal-parquet 
There are some parquet files attached to tickets in jira, too.
Do you plan to convert those into tests?

Regarding FIXED_LENGTH_BYTE_ARRAY.... The overhead would decreases compared to size. BINARY overhead would be <10% from ~DECIMAL(100) and <25% from ~DECIAL(40) (pre-compression). I'd expect DECIMAL(40) to use the full precision only from time to time. But yeah, I've overlooked the 4 byte overhead at https://github.com/Parquet/parquet-format/blob/master/Encodings.md and assumed it would be less, FIXED_LENGTH_BYTE_ARRAY should be good for now (until s.o. complains).

liancheng commented:
@rtreffer Cool thanks! Then I'm merging #6617 shortly. And yes, I plan to convert those Parquet files into tests, probably just include them as test resources.

liancheng commented:
@rtreffer I've merged #6617.

rtreffer committed:
SPARK-4176 spark should save decimal values with precision >18 as parquets fixed_byte_array

Parquet defines multiple ways to store decimals. This patch enables
the reading of all variations as well as writing decimals in the
smallest fixed-length container possible (INT32, INT64, FIXED_LEN_BYTE_ARRAY).

rtreffer commented:
Currently reworking the patch.

Here is the warning about the tuple match
```
[warn] /home/rtreffer/work/spark-master/sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala:334: object Fixed expects 2 patterns to hold (Int, Int) but crushing into 2-tuple to fit single pattern (SI-6675)
```
According to the ticket it's a deprecation warning.
https://issues.scala-lang.org/browse/SI-6675

Nothing urgent, but I think it should be fixed at some point.

sujkh85 commented:
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-4176][WIP] Support decimal types with precision > 18 in parquet (#6796)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35855 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35855/consoleFull) for   PR 6796 at commit [`5fe321e`](https://github.com/apache/spark/commit/5fe321ee027570eea49869bcbe80c55246538229).

rtreffer commented:
@liancheng it starts to work (compiles and minimal initial test worked, no guarantees). I think there are some points that need feedback
- How is the compatibility mode intended to work? Settings are currently private, but I'd like to store Decimal(19), so is lifting the 18 limit correct for compatibility mode?
- INT32/INT64 are only used when the byte length matches the byte length for the precision. FIXED_LEN_BYTE_ARRAY will thus e.g. be used to store 6 byte values
- FIXED_LEN_BYTE_ARRAY means I'll have to create an array of the correct size. I've increased the scratch_bytes. Not very happy about the code path, do you have better ideas?
- BYTES_FOR_PRECISION needs to handle any precision. I've reworked that code. Again, suggestions welcome

The patch is now way smaller and less intrusive. Looks like the refactoring was well worth the effort!

SparkQA commented:
[Test build #35855 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35855/console) for   PR 6796 at commit [`5fe321e`](https://github.com/apache/spark/commit/5fe321ee027570eea49869bcbe80c55246538229).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/CatalystSchemaConverter.scala:
```diff
@@ -383,20 +386,14 @@ private[parquet] class CatalystSchemaConverter(
-            s"When ${SQLConf.PARQUET_FOLLOW_PARQUET_FORMAT_SPEC.key} is set to false," +
-            "decimal precision and scale must be specified, " +
-            "and precision must be less than or equal to 18.")
-
```
We still need this branch to handle the case where precision information is missing.

rtreffer replied:
You said it should usevthe hive default of (10,0) - or did I misinterpret that?

Am 28. Juni 2015 10:53:00 MESZ, schrieb Cheng Lian <notifications@github.com>:
>> @@ -383,20 +386,14 @@ private[parquet] class CatalystSchemaConverter(
>>            .length(minBytesForPrecision(precision))
>>            .named(field.name)
>>  
>> -      case dec @ DecimalType() if !followParquetFormatSpec =>
>> -        throw new AnalysisException(
>> -          s"Data type $dec is not supported. " +
>> -            s"When ${SQLConf.PARQUET_FOLLOW_PARQUET_FORMAT_SPEC.key}
>is set to false," +
>> -            "decimal precision and scale must be specified, " +
>> -            "and precision must be less than or equal to 18.")
>> -
>
>We still need this branch to handle the case where precision
>information is missing.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/apache/spark/pull/6796/files#r33420570

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

liancheng replied:
(Please see my comments [here] [1].)

[1]: https://github.com/apache/spark/pull/6796/files#diff-83ef4d5f1029c8bebb49a0c139fa3154R301

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/CatalystSchemaConverter.scala:
```diff
@@ -169,11 +169,12 @@ private[parquet] class CatalystSchemaConverter(
+          case _ if assumeInt96IsTimestamp => TimestampType
+          case null => makeDecimalType(maxPrecisionForBytes(12))
+          case _ => illegalType()
+        }
```
`INT96` is only used for nanosecond timestamp types for historical reasons, and is to be deprecated. Let's not use it for decimals.

rtreffer replied:
Didn't know about the deprecation, will drop it.

Am 28. Juni 2015 10:56:00 MESZ, schrieb Cheng Lian <notifications@github.com>:
>> @@ -169,11 +169,12 @@ private[parquet] class CatalystSchemaConverter(
>>          }
>>  
>>        case INT96 =>
>> -        CatalystSchemaConverter.analysisRequire(
>> -          assumeInt96IsTimestamp,
>> -          "INT96 is not supported unless it's interpreted as
>timestamp. " +
>> -            s"Please try to set
>${SQLConf.PARQUET_INT96_AS_TIMESTAMP.key} to true.")
>> -        TimestampType
>> +        field.getOriginalType match {
>> +          case DECIMAL => makeDecimalType(maxPrecisionForBytes(12))
>> +          case _ if assumeInt96IsTimestamp => TimestampType
>> +          case null => makeDecimalType(maxPrecisionForBytes(12))
>> +          case _ => illegalType()
>> +        }
>
>`INT96` is only used for nanosecond timestamp types for historical
>reasons, and is to be deprecated. Let's not use it for decimals.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/apache/spark/pull/6796/files#r33420609

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

liancheng replied:
Yeah, it's not mentioned anywhere, just got this information from Parquet dev mailing list :)

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/CatalystSchemaConverter.scala:
```diff
@@ -373,8 +374,10 @@ private[parquet] class CatalystSchemaConverter(
+      // precisions > 18.
       case DecimalType.Fixed(precision, scale)
-        if precision <= maxPrecisionForBytes(8) && !followParquetFormatSpec =>
+        if !followParquetFormatSpec =>
```
Nit: Let's join this line and the line above.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/CatalystSchemaConverter.scala:
```diff
@@ -383,20 +386,14 @@ private[parquet] class CatalystSchemaConverter(
       case DecimalType.Fixed(precision, scale)
-        if precision <= maxPrecisionForBytes(4) && followParquetFormatSpec =>
+        if followParquetFormatSpec && maxPrecisionForBytes(3) < precision && 
+          precision <= maxPrecisionForBytes(4) =>
```
Why do we want `maxPrecisionForBytes(3) < precision` here? Did I miss something?

rtreffer replied:
We had a debate about using the most compact storage type if possible.

As such INT32 looses compared to a 3 byte fixed length array.

Am 28. Juni 2015 10:59:15 MESZ, schrieb Cheng Lian <notifications@github.com>:
>>        case DecimalType.Fixed(precision, scale)
>> -        if precision <= maxPrecisionForBytes(4) &&
>followParquetFormatSpec =>
>> +        if followParquetFormatSpec && maxPrecisionForBytes(3) <
>precision && 
>> +          precision <= maxPrecisionForBytes(4) =>
>
>Why do we want `maxPrecisionForBytes(3) < precision` here? Did I miss
>something?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/apache/spark/pull/6796/files#r33420647

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

liancheng replied:
(Please see my comment [here] [1].)

[1]: https://github.com/apache/spark/pull/6796#issuecomment-116766442

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/CatalystSchemaConverter.scala:
```diff
@@ -404,9 +401,10 @@ private[parquet] class CatalystSchemaConverter(
       case DecimalType.Fixed(precision, scale)
-        if precision <= maxPrecisionForBytes(8) && followParquetFormatSpec =>
+        if followParquetFormatSpec && maxPrecisionForBytes(7) < precision &&
+          precision <= maxPrecisionForBytes(8) =>
```
Same question as above...

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/CatalystSchemaConverter.scala:
```diff
@@ -562,4 +560,5 @@ private[parquet] object CatalystSchemaConverter {
       throw new AnalysisException(message)
     }
   }
+
```
Nit: Remove this newline.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableSupport.scala:
```diff
@@ -212,10 +212,7 @@ private[parquet] class RowWriteSupport extends WriteSupport[InternalRow] with Lo
-            sys.error(s"Unsupported datatype $d, cannot write to consumer")
-          }
-          writeDecimal(value.asInstanceOf[Decimal], d.precisionInfo.get.precision)
+          writeDecimal(value.asInstanceOf[Decimal], d.precisionInfo.map(_.precision).getOrElse(10))
```
Need to report error for `DecimalType(None)`.

liancheng commented:
@rtreffer Thanks for rebasing and simplifying this! I left some comments but haven't finished my review, will be back after confirming some details related to your questions.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala:
```diff
@@ -43,16 +43,27 @@ private[parquet] object ParquetTypesConverter extends Logging {
   }
 
+  private[parquet] def BYTES_FOR_PRECISION_STATIC =
+    (0 to 30).map(BYTES_FOR_PRECISION_COMPUTE).toArray
```
30 should probably be replaced with 38, which fits in 16 bytes, and is the maximum precision supported in Hive.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala:
```diff
@@ -43,16 +43,27 @@ private[parquet] object ParquetTypesConverter extends Logging {
    */
-  private[parquet] val BYTES_FOR_PRECISION = Array.tabulate[Int](38) { precision =>
-    var length = 1
+  private[parquet] def BYTES_FOR_PRECISION_COMPUTE(precision : Int) : Int = {
```
Prefer `bytesForPrecision` since it's a method instead of a constant.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala:
```diff
@@ -43,16 +43,27 @@ private[parquet] object ParquetTypesConverter extends Logging {
     length
   }
 
+  private[parquet] def BYTES_FOR_PRECISION_STATIC =
```
Prefer `bytesForPrecisionStatic`.

liancheng commented:
Hi @rtreffer,

> - How is the compatibility mode intended to work? Settings are currently private, but I'd like to store Decimal(19), so is lifting the 18 limit correct for compatibility mode?

The compatibility mode is enabled by setting `spark.sql.parquet.followParquetFormatSpec` to `false`.  This mode must be enabled for now, because the write path hasn't been refactored to follow the Parquet format spec.  Note that compatibility mode only affects the write path, because the Parquet format spec also covers legacy formats by various backwards-compatibilty rules.

Decimals with precision > 18 could be enabled even in compatibility mode.  Because it doesn't affect compatibility: old Spark versions can't read decimals with precision > 18 from the very beginning.

What do you mean by saying "settings are currently private"?  `SQLConf.PARQUET_FOLLOW_PARQUET_FORMAT_SPEC` is `private[spark]`, all classes under `org.apache.spark` can access it.

> - INT32/INT64 are only used when the byte length matches the byte length for the precision. FIXED_LEN_BYTE_ARRAY will thus e.g. be used to store 6 byte values

I see your point.  You mentioned a "debate" in [this comment] [1], were you referring to [this one] [2]?  From the perspective of storage efficiency, it probably makes sense.  (I said "probably" because I'm not quite sure about the average case after taking encoding/compression into consideration.)  However, in the case of Parquet, we usually care more about speed and memory consumption.  Especially, Parquet can be super memory consuming when reading files with wide schema (i.e., large column number).  A key advantage of `INT32` and `INT64` is that, they avoid boxing costs in many cases and thus can be faster and use less memory.  Also, you don't need to do all those bit operations to encode/decode the unscaled long value of a decimal when using `INT32` and `INT64`.

At the meantime, Parquet handles `INT32` and `INT64` pretty efficiently.  There are more encoders for integral types than binaries (either fixed-length or not, see [Encodings.md] [3] for more details).  Although I haven't done benchmark for this, but I believe in many cases, storage efficiency of `INT32` can be comparable or even better than `FIXED_LEN_BYTE_ARRAY` with a length less than 4. The same should also applies to `INT64`.

So I suggest: when compatibility mode is off, we just use `INT32` for 1 <= precision <= 9, and `INT64` for 10 <= precision <= 18 when converting `DecimalType`s in `CatalystSchemaConverter`.  When we refactor the write path to follow Parquet format spec, we can write decimals in `INT32` and `INT64` when appropriate in follow-up PRs.

The TL;DR is: I'd just remove `precision <= maxPrecisionForBytes(8)` in [this line] [4] and leave everything else unmodified (you comment updates looks good to me though :)

> - FIXED_LEN_BYTE_ARRAY means I'll have to create an array of the correct size. I've increased the scratch_bytes. Not very happy about the code path, do you have better ideas?

Hive limits the max precision of a decimal to 38, which fits in 16 bytes.  So 16 rather than 4096 bytes should be enough for most cases.  Also it would be better to refactor branches of [this `if` expression] [5] into two separate methods for clarity.  Otherwise it looks good.

> - BYTES_FOR_PRECISION needs to handle any precision. I've reworked that code. Again, suggestions welcome

(See my other comments inlined.)

[1]: https://github.com/apache/spark/pull/6796#discussion_r33420742
[2]: https://github.com/apache/spark/pull/6796#discussion_r32891515
[3]: https://github.com/Parquet/parquet-format/blob/master/Encodings.md
[4]: https://github.com/apache/spark/pull/6796/files#diff-a4c01298c63223d113645a31c01141baL377
[5]: https://github.com/apache/spark/pull/6796/files#diff-83ef4d5f1029c8bebb49a0c139fa3154R301

rtreffer commented:
Hi @liancheng, thank you for the thorough review, will push a reworked version soon. Everything sounds reasonable :-)

With "private" Settings I meant that I can't change the setting in the shell because it's marked as "isPublic = false" in https://github.com/liancheng/spark/blob/2a2062d3f530ecd26e75b306aee42761d67d8724/sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala#L273

I'm not sure if that's intended.

liancheng commented:
@rtreffer Yeah, it's intended. As explained above, this feature flag must be set to `false` for now because the write path hasn't been refactored to respect the Parquet format spec. If we turn this on, `CatalystSchemaConverter` will generate standard Parquet schema while the write path still writes data conforming to the old legacy format, which leads to data corruption.

liancheng commented:
Hey @rtreffer, just want to make sure whether you are still working on this? I'm asking because I just opened #7231 to refactor Parquet read path for interoperability and backwards-compatibility, which also touches the decimal parts. I believe the new [`CatalystDecimalConverter`] [1] already covers the read path of decimals with precision > 18, which means this PR can be further simplified. Just in case you don't have time to continue this PR, I'm happy to fork your branch and get it merged (will still list you as the main author).

[1]: https://github.com/apache/spark/pull/7231/files#diff-1d6c363c04155a9328fe1f5bd08a2f90R237

rtreffer commented:
Hi @liancheng,

I'm rebasing on you PR right now. I can work for ~1-2h / day on this PR so feel free to take over the PR if this blocks anything.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36702 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36702/consoleFull) for   PR 6796 at commit [`e6dad45`](https://github.com/apache/spark/commit/e6dad4574f47a7b6694500df2a1b860555537c86).

SparkQA commented:
[Test build #36702 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36702/console) for   PR 6796 at commit [`e6dad45`](https://github.com/apache/spark/commit/e6dad4574f47a7b6694500df2a1b860555537c86).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36703 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36703/consoleFull) for   PR 6796 at commit [`7a57c16`](https://github.com/apache/spark/commit/7a57c163ec3fe516d3b173042329a9b6b135efa9).

rtreffer commented:
The writeDecimal method is rather ugly, and the write path needs to know if we follow parquet style or not as this implies a different encoding (addInteger / addLong).

SparkQA commented:
[Test build #36703 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36703/console) for   PR 6796 at commit [`7a57c16`](https://github.com/apache/spark/commit/7a57c163ec3fe516d3b173042329a9b6b135efa9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37097 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37097/consoleFull) for   PR 6796 at commit [`1152721`](https://github.com/apache/spark/commit/1152721ebeafe7a4535e202c3091f415a8ba3863).

SparkQA commented:
[Test build #37097 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37097/console) for   PR 6796 at commit [`1152721`](https://github.com/apache/spark/commit/1152721ebeafe7a4535e202c3091f415a8ba3863).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37099 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37099/consoleFull) for   PR 6796 at commit [`3e30bdf`](https://github.com/apache/spark/commit/3e30bdfb1199a105a882dde7d2dc0bd8edea05a2).

SparkQA commented:
[Test build #37099 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37099/console) for   PR 6796 at commit [`3e30bdf`](https://github.com/apache/spark/commit/3e30bdfb1199a105a882dde7d2dc0bd8edea05a2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1055 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1055/consoleFull) for   PR 6796 at commit [`3e30bdf`](https://github.com/apache/spark/commit/3e30bdfb1199a105a882dde7d2dc0bd8edea05a2).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37130 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37130/consoleFull) for   PR 6796 at commit [`c8d4d6c`](https://github.com/apache/spark/commit/c8d4d6c9f0e420b2bd54e358b6b73f198ef3373e).

SparkQA commented:
[Test build #37130 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37130/console) for   PR 6796 at commit [`c8d4d6c`](https://github.com/apache/spark/commit/c8d4d6c9f0e420b2bd54e358b6b73f198ef3373e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1055 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1055/console) for   PR 6796 at commit [`3e30bdf`](https://github.com/apache/spark/commit/3e30bdfb1199a105a882dde7d2dc0bd8edea05a2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37143 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37143/consoleFull) for   PR 6796 at commit [`1703c26`](https://github.com/apache/spark/commit/1703c26f917c5e06f60bc9c8cd9299c9ffbb2389).

SparkQA commented:
[Test build #37143 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37143/console) for   PR 6796 at commit [`1703c26`](https://github.com/apache/spark/commit/1703c26f917c5e06f60bc9c8cd9299c9ffbb2389).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37146 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37146/consoleFull) for   PR 6796 at commit [`83ca029`](https://github.com/apache/spark/commit/83ca029b2ec6e940f73acf9da0eae34319baeb6b).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37147 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37147/consoleFull) for   PR 6796 at commit [`1dad677`](https://github.com/apache/spark/commit/1dad677449445c878d7d938192df4a6b2d997db4).

SparkQA commented:
[Test build #37146 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37146/console) for   PR 6796 at commit [`83ca029`](https://github.com/apache/spark/commit/83ca029b2ec6e940f73acf9da0eae34319baeb6b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #37147 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37147/console) for   PR 6796 at commit [`1dad677`](https://github.com/apache/spark/commit/1dad677449445c878d7d938192df4a6b2d997db4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

liancheng commented:
@rtreffer Since #7455 supersedes this PR, would you mind to close this one?

rtreffer commented:
@liancheng sure, I just wasn't sure if it should be closed :-)
